### PR TITLE
cleanup associations when handshake didn't complete, #29615

### PR DESCRIPTION
**Pull request by patriknw**
* symptom was continously logged "Quarantine of [] ignored because unknown UID,
  triggered from the Association.setupIdleTimer task
* when handshake wasn't completed for some reason the UID is unknown and then
  this task continues forever
* abort such associations and cancel the idle task in that case
* also cleanup such associations from AssociationRegistry.removeUnusedQuarantined

References #29615

I tested with an example, and had to simulate broken handshake exchange.

### CODE REVIEW

**Review by octonato — COMMENTED** [review]
I find it hard to reason about when we juggle with two booleans that are related. `isQuarantined` depends on the address, which is also used in `isHandshakeCompleted`. 

We have three cases, that I guess, are relevant for the decisions we are making:

1. we DO NOT have the remote address
1. we have the remote address and its UID IS on the quarantine map
1. we have the remote address and its UID IS NOT on the quarantine map

Then we use `isQuarantined` and `isHandshakeCompleted` with boolean inversion `!` and `&&` and `||` in order to know if we are in one of the three situations above. 

Maybe we should model it with a three state ADT.
```scala
sealed trait AddressState
case object AddressUnknow extends AddressState
case object AddressQuarantined extends AddressState
case object AddressWithMissingUID extends AddressState
```
Last name sucks, we can find something better. 

Then, when using it inside `if`s we will express more clearly where we are and make it easier to understand why we are taking this or that decision.

File: akka-remote/src/main/scala/akka/remote/artery/Association.scala [hunk 1]
```diff
@@ -666,6 +666,12 @@ private[remote] class Association(
             // If idle longer than quarantine-idle-outbound-after and the low frequency HandshakeReq
             // doesn't get through it will be quarantined to cleanup lingering associations to crashed systems.
             quarantine(s"Idle longer than quarantine-idle-outbound-after [${QuarantineIdleOutboundAfter.pretty}]")
+            val newLastUsedDurationNanos = System.nanoTime() - associationState.lastUsedTimestamp.get
+            if (!associationState.isQuarantined() && !associationState
+                  .isHandshakeCompleted() && newLastUsedDurationNanos >= QuarantineIdleOutboundAfter.toNanos) {
+              // quarantine ignored due to unknown UID, have to stop this task anyway
+              abortQuarantined()
```
[the remark this answers is not in the document]
Comment (octonato) [reply akka-remote/src/main/scala/akka/remote/artery/Association.scala:673]: if what I describe above is correct and if we have a three state ADT, that code could be written as:

```scala
if (associationState.addressStae == AddressWithMissingUID 
	&& newLastUsedDurationNanos >= QuarantineIdleOutboundAfter.toNanos) {
	// quarantine ignored due to unknown UID, have to stop this task anyway
   abortQuarantined()
}
```

File: akka-remote/src/main/scala/akka/remote/artery/Association.scala [hunk 2]
```diff
@@ -1166,7 +1172,8 @@ private[remote] class AssociationRegistry(createAssociation: Address => Associat
     val remove = currentMap.foldLeft(Map.empty[Address, Association]) {
       case (acc, (address, association)) =>
         val state = association.associationState
-        if (state.isQuarantined() && ((now - state.lastUsedTimestamp.get) >= afterNanos))
+        if ((state.isQuarantined() || !state
+              .isHandshakeCompleted()) && ((now - state.lastUsedTimestamp.get) >= afterNanos))
```
Comment (octonato) [inline akka-remote/src/main/scala/akka/remote/artery/Association.scala:1176]: With three state ADT, this one would be:

```scala
if (state.addressState != AddressWithMissingUID 
    && ((now - state.lastUsedTimestamp.get) >= afterNanos))
```

File: akka-remote/src/main/scala/akka/remote/artery/Association.scala [hunk 3]
```diff
@@ -1188,7 +1195,8 @@ private[remote] class AssociationRegistry(createAssociation: Address => Associat
     currentMap.keysIterator.foreach { uid =>
       val association = currentMap.get(uid).get
       val state = association.associationState
-      if (state.isQuarantined() && ((now - state.lastUsedTimestamp.get) >= afterNanos))
+      if ((state.isQuarantined() || !state
+            .isHandshakeCompleted()) && ((now - state.lastUsedTimestamp.get) >= afterNanos))
```
Comment (octonato) [inline akka-remote/src/main/scala/akka/remote/artery/Association.scala:1199]: same as previous:

```scala
if (state.addressState != AddressWithMissingUID 
    && ((now - state.lastUsedTimestamp.get) >= afterNanos))
```

Maybe we don't even need a three state. All that we are trying to detected is if we are in the case of having an address but missing its UID. Maybe once single method with a clear name will do the trick.

**Comment by patriknw** [comment]
@octonato I have added the 3 state ADT. Take a look when you have time ad705b9dc3
After that I will have to test things again with my simulated example.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 4]
```diff
@@ -106,6 +106,12 @@ private[remote] object AssociationState {
   private final case class UniqueRemoteAddressValue(
       uniqueRemoteAddress: Option[UniqueAddress],
       listeners: List[UniqueAddress => Unit])
+
+  sealed trait UniqueRemoteAddressState
+  final case class UidKnown(uniqueAddress: UniqueAddress) extends UniqueRemoteAddressState
+  case object UidUnknown extends UniqueRemoteAddressState
+  final case class UidQuarantined(uniqueAddress: UniqueAddress) extends UniqueRemoteAddressState
```
Comment (octonato) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:113]: I think they may all be case objects. Whenever we pattern match on them, we are using `(_)` for the case classes. We simply don't use the value at all.

File: akka-remote/src/main/scala/akka/remote/artery/Association.scala [hunk 5]
```diff
@@ -666,6 +666,15 @@ private[remote] class Association(
             // If idle longer than quarantine-idle-outbound-after and the low frequency HandshakeReq
             // doesn't get through it will be quarantined to cleanup lingering associations to crashed systems.
             quarantine(s"Idle longer than quarantine-idle-outbound-after [${QuarantineIdleOutboundAfter.pretty}]")
+            associationState.uniqueRemoteAddressState() match {
+              case AssociationState.UidQuarantined(_) => // quarantined as expected
+              case AssociationState.UidKnown(_)       => // must be new uid, keep as is
```
Comment (octonato) [inline akka-remote/src/main/scala/akka/remote/artery/Association.scala:671]: can be:

```scala 
case AssociationState.UidQuarantined | AssociationState.UidKnown => 
```

(assuming we make them all case objects)

**Review by octonato — APPROVED** [review]
Thanks @patriknw. 

This is looking much better. 

I think we can make them all case object, see comments.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala (hunk 4 above)
Comment (patriknw) [reply akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:113]: um, I guess I was thinking of other imaginary usages of these. I'll change to objects.

**Review by johanandren — APPROVED** [review]
LGTM but Renatos suggestions sounds good.

**Review by johanandren — APPROVED** [review]

**Comment by johanandren** [comment]
Is this waiting for some repeat job @patriknw ? If not I think we can merge it.

**Pull request merged**